### PR TITLE
fuel_gauge: Fix sbs_gauge err to conform to API

### PR DESCRIPTION
**Pull request by aaronemassey**
The SBS fuel gauge driver did not return a count of the number of properties that failed.

Fix this and add a test that verifies the sbs_gauge returns the number of failing properties as well as a test that verifies a negative return code if all properties failed.

Signed-off-by: Aaron Massey <aaronmassey@google.com>

### CODE REVIEW

**Review by alevkoy — APPROVED** [review]
In the spirit of incremental improvement, this change seems fine.

But I think separate return values for some-failed vs all-failed is an odd convention. The fact that positive return codes were broken before would imply that no one is using them correctly now. I would suggest just returning a negative value (maybe the same value) for any number of failures, and the caller can inspect the property array if they care about the details, which they seem to need to do anyway.

File: drivers/fuel_gauge/sbs_gauge/sbs_gauge.c [hunk 1]
```diff
@@ -93,13 +93,17 @@ static int sbs_gauge_get_prop(const struct device *dev, struct fuel_gauge_get_pr
...
 
 	for (int i = 0; i < len; i++) {
-		ret |= sbs_gauge_get_prop(dev, props + i);
+		int ret = sbs_gauge_get_prop(dev, props + i);
+
+		err_count += ret ? 1 : 0;
 	}
 
-	return ret;
+	err_count = (err_count == len) ? -1 : err_count;
```
[the remark this answers is not in the document]
Comment (aaronemassey) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: > The fact that positive return codes were broken before would imply that no one is using them correctly now. 

No one is using them now :stuck_out_tongue: this is the first experimental driver.

> I would suggest just returning a negative value (maybe the same value) for any number of failures, and the caller can inspect the property array if they care about the details, which they seem to need to do anyway.

I think you make a good point. @yperess thoughts?

[the remark this answers is not in the document]
Comment (rgundi) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: I too think it is better to return only 0 or a negative value. Additional details, if required, can be got from the property array.

[the remark this answers is not in the document]
Comment (aaronemassey) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: @rgundi Happy to change this part of the API.

How do we feel about positive value for number of errors and 0 for no errors. I could do negative value for number of errors too, but it makes more sense to me for it to be a (uint_16).

[the remark this answers is not in the document]
Comment (alevkoy) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: I think we conventionally use negative status indicators for errors based on the type of error. In what scenario is the number of individual errors interesting?

[the remark this answers is not in the document]
Comment (aaronemassey) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: So, this is not a total sticking point for me.

But let's say we're updating the parameters for 20 properties, if some, let's say n, of them failed, I may want to allocate an n-sized array of bad statuses. Maybe I want to report them somewhere or retry getting those n failed properties without iterating over all of the properties. Each time I retry getting them I'd have to iterate over all of them.

But this is also a very specific situation where I'm trying to save a pretty small amount of boilerplate on the client.

[the remark this answers is not in the document]
Comment (yperess) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: I'll argue that we don't need to come up with a reason. The point is that on an error we cannot give any meaningful error code except just "something went wrong". As such, we're left with returning the number of errors. Given the choice between `-num_errs` or `num_errs` I prefer a positive number. It avoids the confusion of trying to read it as a standard error code.

Tldr

You cannot give a meaningful error code, so the count makes the most sense. Positive numbers avoid namespace issues with standard error codes. I vote positive error count.

[the remark this answers is not in the document]
Comment (rgundi) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: hmm.. interesting viewpoints. If you are in favor of using a positive number, wouldn't it be better if you return the number of passing properties? len is anyway being passed and if the return value doesn't match len, it means some property has failed.

[the remark this answers is not in the document]
Comment (alevkoy) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: > But let's say we're updating the parameters for 20 properties, if some, let's say n, of them failed, I may want to allocate an n-sized array of bad statuses. Maybe I want to report them somewhere or retry getting those n failed properties without iterating over all of the properties. Each time I retry getting them I'd have to iterate over all of them.

I don't want to do dive too deeply into those examples, since I think they were just hypothetical, but the reason I don't think this is that useful is that any non-trivial error handling is going to need to iterate over the array _anyway_, and the time taken by is going to be dominated by retrying the I2C/SPI transactions _anyway_, and that applies to the scenarios you proposed. At a higher level, we really don't expect any of these transactions to fail in a normal system. Even if we could meaningfully optimize here, I doubt that it's worth it.

> You cannot give a meaningful error code, so the count makes the most sense.

Basically, I disagree. I think we should make the API as simple as possible without a compelling reason to make it more complicated. If we can't return meaningful error codes, then `bool` seems like a reasonable choice to me. (If that's unconventional, I'd just pick `-EIO` or some other code for all failures.) I suspect that, in almost every practical case, the most reasonable thing to do if any transaction fails will just be to abandon the whole process and _maybe_ iterate the array to log which reads failed. I wouldn't want to give users the impression that we think the number of errors is important and should be used typically.

To elaborate on simple-vs-useful slightly, I started to write that we should just return `void` if our error codes aren't meaningful. That wouldn't lose any information. But it would require the caller to iterate over the array to tell if all the reads succeeded, which would make the most likely code path slower and more complicated. I think making the API slightly more complicated/redundant by returning `bool` instead of `void` is clearly worth it.

[the remark this answers is not in the document]
Comment (aaronemassey) [reply drivers/fuel_gauge/sbs_gauge/sbs_gauge.c:104]: Moving discussion to https://github.com/zephyrproject-rtos/zephyr/issues/52992.

**Review by rgundi — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
